### PR TITLE
[libSyntax] Create SyntaxRef, which uses SyntaxDataRef internally

### PR DESCRIPTION
**Pull request by ahoppen**
Now that we have a fast `SyntaxDataRef`, create a corresponding `SyntaxRef` hierarchy. In contrast to the `Syntax` hierarchy, `SyntaxRef` does **not** own the backing `SyntaxDataRef`, but merely has a pointer to the `SyntaxDataRef`. In addition to the requirements imposed by `SyntaxDataRef`, the user of the `SyntaxRef` hierarchy needs to make sure that the backing `SyntaxDataRef` of a `SyntaxRef` node stays alive. While this sounds like a lot of requirements, it has performance advantages:
 - Passing a `SyntaxRef` node around is just passing a pointer around.
 - When casting a `SyntaxRef` node, we only need to create a new `SyntaxRef` (aka. pointer) that points to the same underlying `SyntaxDataRef` – there's no need to duplicate the `SyntaxDataRef`.
 - As `SyntaxDataRef` is not ref-counted, there's no ref-counting overhead involved.

Furthermore, the requirements are typically fulfilled. The `getChild` methods on SyntaxRef return an `OwnedSyntaxRef`, which stores the `SyntaxDataRef`. As long as this variable is stored somewhere on the stack, the corresponding `SyntaxRef` can safely be used for the duration of the stack frame. Even calls like the following are possible, because `OwnedSyntaxRef` returned by `getChild` stays alive for the duration of the entire statement.

```cpp
useSyntaxRef(mySyntaxRef.getChild(0).getRef())
```

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Review by rintaro — COMMENTED** [review]
Overall looks good.

File: include/swift/Syntax/Syntax.h [hunk 1]
```diff
@@ -47,19 +63,137 @@ SyntaxNode makeRoot(const RawSyntax *Raw) {
...
 class Syntax {
 protected:
   RC<const SyntaxData> Data;
 
 public:
-  explicit Syntax(const RC<const SyntaxData> &Data) : Data(Data) {}
-
-  virtual ~Syntax() {}
+  explicit Syntax(const RC<const SyntaxData> &Data) : Data(Data) {
+    assert(Data != nullptr && "SyntaxData must be backed by Data");
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply include/swift/Syntax/Syntax.h:195]: Yes

File: include/swift/Syntax/Syntax.h [hunk 2]
```diff
@@ -47,19 +63,137 @@ SyntaxNode makeRoot(const RawSyntax *Raw) {
...
+
+/// See comment on top of file.
 class Syntax {
 protected:
   RC<const SyntaxData> Data;
 
 public:
-  explicit Syntax(const RC<const SyntaxData> &Data) : Data(Data) {}
-
-  virtual ~Syntax() {}
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply include/swift/Syntax/Syntax.h:62]: I *guess* that the specific `Syntax` nodes had their own destructors at some point and that they went away during some change (none of my recent ones anyway). I removed it, nothing broke, so I assumed it’s not necessary and removed it just to clear things up. There’s no specific reason to keep or remove it.

File: include/swift/Syntax/Syntax.h [hunk 3]
```diff
@@ -194,8 +328,196 @@ class Syntax {
   AbsoluteOffsetPosition getAbsoluteEndPositionAfterTrailingTrivia() const {
     return Data->getAbsoluteEndPositionAfterTrailingTrivia();
   }
+};
+
+// MARK: - SyntaxRef
+
+/// See comment on top of file.
+class SyntaxRef {
+  const SyntaxDataRef *Data;
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply include/swift/Syntax/Syntax.h:337]: As discussed offline, I’ll set up a follow-up PR to change the name of `SyntaxDataRef`. On there, we can also discuss what a better name might be.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by Bareypatch** [comment]
Merge

File: include/swift/Syntax/Syntax.h (hunk 3 above)
Comment (rintaro) [inline include/swift/Syntax/Syntax.h:337]: I think the pointer itself is also never changed. So `const SyntaxDataRef * const Data`

File: include/swift/Syntax/Syntax.h [hunk 4]
```diff
@@ -47,19 +63,137 @@ SyntaxNode makeRoot(const RawSyntax *Raw) {
...
+      : Data(std::move(Other.Data)), Ref(getDataPtr(), no_validation_t()) {}
+
+  /// The pointer to the location at which \c this stores the \c Data.
+  /// Can be used to retroactively populate the \c Data after \c OwnedSyntaxRef
+  /// has been constructed with uninitialised memory.
+  SyntaxDataRef *getDataPtr() { return &Data; }
+
+  const SyntaxRefType &getRef() {
+    assert(Ref.getDataRef() == getDataPtr() &&
+           "Ref no longer pointing to Data?");
```
Comment (rintaro) [inline include/swift/Syntax/Syntax.h:121]: If `Data` is const, this `assert` is not needed.

Comment (ahoppen) [reply include/swift/Syntax/Syntax.h:121]: I ran into issues here at some point after moving `OwnedSyntaxRef` to a different location. The pointer stays constant but the `SyntaxDataRef` now lives in a different location, so the pointer isn’t valid anymore. That’s why I put the assertion in and I think it’s still useful if someone starts to fiddle around `OwnedSyntaxRef`.

File: include/swift/Syntax/Syntax.h (hunk 3 above)
Comment (ahoppen) [reply include/swift/Syntax/Syntax.h]: 👍 Also changed `Data` in `Syntax` to `const RC<const SyntaxData> Data`.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Review by rintaro — APPROVED** [review]

**Pull request merged**